### ISSUE TITLE
Port build system to Bazel

### ISSUE THREAD

**Issue by per-gron** [zcash/zcash#2811]
Would you be interested in accepting PRs that move to/towards building ZCash with Bazel?

I have seen #1163, which was closed due to "favoring gitian instead" but it provides no reasoning as to why Gitian is preferred (was it a compromise because using it was easier?). Benefits that a Bazel-based build system could provide include:

* Support macOS and Windows
* Reproducible builds, on all platforms (only if configured correctly, but unlike most other build systems it is actually feasible with Bazel)
* Simpler build process: There is no need for a special mode for deterministic release builds, cross-compilation etc
* Nicer caching than ccache

I'm asking because for me personally getting zcash to build was a bit of a pain, and looking at the Github issues it seems like there are plenty of build issues that are much easier to fix with Bazel (or some other nice build tool) than with the current build system.

I'm proposing to use Bazel because I think deterministic builds are extra valuable to zcash due to security requirements, and Bazel is the most well-supported build system that I know of that does that well.

**Comment by str4d** [comment]
> was it a compromise because using it was easier?

Yes; the majority of the codebase already built with Gitian, so all we had to do was introduce our modifications.

Assigning @nathan-at-least for overall technical direction. Personal concept ACK, but I am concerned that replacing the existing build system entirely will be a very disruptive change, particularly for backporting PRs from upstream. Is it possible to implement a Bazel build system in parallel, leaving the existing build system intact? I have done this in other projects previously with Gradle, and it's a nice migration strategy. Once at a minimally-working level, we could have a single CI worker using it to ensure it stays working as upstream / other changes are made.

**Comment by per-gron** [comment]
This is good feedback :) I'm using it at work and have been tinkering with Bazel+zcash a bit and here are some more thoughts:

Bazel is an opinionated build system. To successfully use it, you have to just let it do things its own way, at least to some extent. Many of its idiosyncrasies are carefully thought out and not random. For example: The unit testing targets tend to end up linking to a surprisingly large number of shared libraries. (It is possible to avoid that through configuration but then you get large binaries and/or slow linking instead.) I'll list some more examples below.

Even then, there are a few important design choices that have to be made:

1. **Deterministic vs (somewhat) lightweight builds**: By default, Bazel uses the compiler toolchain that is installed on the system, but it provides good tools for making the build always download a specific toolchain and use that. **I think Zcash would benefit from deterministic builds, where the build downloads and uses its own specific version of Rust and (at least for now) GCC on all platforms**, both for security reasons and because there is little reason to spend time supporting many versions of different compilers and various combinations of library dependencies.
2. **Granularity**: It is possible to use globbing (ie `**/*.cpp`) in the BUILD files to keep them very short, but many of Bazel's advanced features like caching, (potentially) distributed execution etc don't work that well then. Bazel is optimized for having basically every `.cpp` file as its own "library". When used like that, Bazel requires all dependencies between all source files to be written down explicitly in BUILD files (this is why you get lots of shared libraries for unit tests). BUILD files become fairly verbose then, and Bazel strictly enforces that there are no circular dependencies. **I would recommend specifying the build with as high granularity as possible** (ie group files together when there are circular dependencies), to get the most out of it. But this will also add a certain amount of bookkeeping work to day-to-day development. It does help maintain good structure and tooling though. It is ok to not go all the way with granularity though.
3. **Header search paths**: Bazel is relatively flexible wrt header search paths, but it definitely nudges you towards having only one header search path for all code in the whole project (except for external dependencies), monorepo style. It specifically does not easily support use of system-style includes for non-external headers (with `<>` instead of quotes). **I suggest changing the code base to not use `<>` style includes for headers in the zcash source tree (it's clearer that way) but otherwise keep the current header search paths for now.**
4. **Build time flags**: The idea of having a pre-compile configure step that figures out what libraries you want to use is somewhat at odds with deterministic builds where the build system downloads everything it needs in the exact version it wants. **I suggest removing all configure-time options from the build and using only runtime configuration.** This is the quickest thing to do to get the build up and running and also I believe superior in general.

> I am concerned that replacing the existing build system entirely will be a very disruptive change, particularly for backporting PRs from upstream. Is it possible to implement a Bazel build system in parallel, leaving the existing build system intact?

Yes that should be doable. I agree that doing so makes this more likely to succeed. For the most part I think the initial version of a Bazel based build system will not change the source files much (some `#include <...>` might be changed to `#include "..."` but not much more). It can technically cause trouble with backporting PRs from upstream though: For example, if an upstream change introduces a long cycle in dependencies between files it could require a large refactoring of either the BUILD files or the underlying code.

In the future, you may want to refactor the code to remove circular dependencies, which would further increase merge conflicts when backporting PRs. Another change that Bazel will encourage is to put the Rust code in the zcash repo.

**Comment by per-gron** [comment]
Quick update: I'm still hacking on this. I am now able to build zcashd, zcash-cli and zcash-tx with Bazel on Linux. The build downloads specific versions of all libraries and the compiler toolchain(s) and does not use any system headers or libraries so should be relatively close to being hermetically sealed and deterministic. There is still a bunch of work left though, including tests, ensuring that the development workflow is good, other build artifacts (libzcashconsensus.so, Debian package, binary tarball etc), some work for librustzcash, not depending on the system's m4, making sure that the GCC that is used in the build can be built in a way that allows reviewers to verify the toolchain tarball hash etc. It would of course also be good to support more than Linux.

Ping @radix42, who is working on [Improved multi-platform support](https://github.com/ZcashFoundation/GrantProposals-2017Q4/files/1364175/CrossPlatformGrantProposal.pdf) which is different but related. Do you use GCC for building zcash on Mac and Windows as well or do you use Clang/MSVC?

**Comment by per-gron** [comment]
Another status update: I am working on adding unit tests now. They build and most of them work but there are 2 or 3 tests that still fail. After that, what is missing are other tests run by `qa/zcash/full_test_suite.py` and a little bit of Rust work.

**Comment by per-gron** [comment]
Current status is that the executables are built and the whole test suite runs and passes, except sec-hard, which I haven't done yet. Next I'm going to tie up some loose ends and try to make something reviewable out of it

**Comment by radix42** [comment]
I just saw this thread, I missed it when I got tagged on the 2nd!

My past Mac ports used gcc from home brew, but I am going to switch to clang now that it builds with it on the latest Xcode. The Mac port by @kozyilmaz uses it, which I’ll be  upstreaming (crosses fingers)

On Windows I’m cross compiling with gcc on Debian

**Comment by per-gron** [comment]
@radix42 Thanks for the info! Once the Linux build is in and I start hacking on Mac support I am leaning towards using gcc there too, simply because I will most likely have to build my own toolchain anyway* and it's just easier to use one compiler everywhere and not have to worry about tiny differences between compilers.

Using different compilers for different platforms is a pain for development and adds little value: it's really not nice to have a working patch just to have it fail in CI due to some obscure warning that is treated as an error only on some platform you don't have easy access to.

Another potential reason for using GCC is that I saw somewhere that clang miscompiles GMP, and since this build setup will compile it from source that could be an issue.

(I personally like clang than GCC though.)

* I don't think it's allowed to have a build script that downloads and uses a specific version of xcode's compiler without showing a license prompt.

**Comment by radix42** [comment]
@kozyilmaz was using strictly gcc before dec 1 on Mac, with a custom self
contained source built gcc toolchain.

**Comment by per-gron** [comment]
I can totally see why you use clang, it is very reasonable. I am looking at custom toolchains because my goal would be that with any Mac with Bazel installed it should be possible to check out the zcash repo at the right commit, do `bazel -c opt build ...` and get a `zcashd` identical to the official release. This is hard to do otherwise.

**Comment by nathan-at-least** [comment]
I'm going to assign @charlieok to explore this (without committing our full packaging pipeline to switch over), because they're exploring a few approaches to windows builds in our CI and packaging pipeline. If this is notably nicer, then the next step would be for us to figure out how painful transitioning away from gitian for all platforms might be.

(Note: I'm not sure where this would fit in the devinfra team's overall priorities; paging @charlieok, @mdr0id, and @mms710.)

**Comment by mms710** [comment]
I'll work with @mdr0id to determine priority of investigating this. Thanks to @per-gron and @radix42  for doing so much great work on this and let us know if we can help support you. I hope to have an update in the next couple weeks on priority.

**Comment by radix42** [comment]
One "gotcha" about Windows builds: when you run zcash-gtest.exe you REALLY want to do so under PowerShell, any other shell and you need to dismiss an error dialog for the one test that is SUPPOSED to fail, after which it keeps on chugging (observed running under cmd.exe and msys2 bash).

also PowerShell is the only one which displays the ASCII/ANSI art when showmetrics=1 when running zcashd.exe.....but cmd.exe is preferred for manual use of zcash-cli.exe, as you do not need to escape quotes AT ALL, EVAR (DOS ftw)

**Comment by charlieok** [comment]
@radix42 interesting... I ran into error dialogs when running some tests (maybe it was `zcash-gtest.exe`) but hadn't considered at the time that maybe the shell I was using was the issue.

Since then I've read through Mircosoft's recent four-part series on their command line blog about how cmd.exe works vs how unix virtual terminals work, and some new features they're adding to new versions of cmd.exe to better bridge those two worlds and continue to support software that uses the old cmd.exe interface.

https://blogs.msdn.microsoft.com/commandline/2018/06/20/windows-command-line-backgrounder/

I'd like to better understand how that relates to these error dialogs, and how powershell fits into this picture. If you happen to know the answers to some of these questions I'd be grateful for any light you can shed...

- Are the error dialogs are the result of an 'error' command that an app sends via the cmd interface?

- When you launch an executable like zcash-gtest.exe with powershell, does powershell use a different (newer) interface to do I/O than cmd does? And would that explain why the error dialog behavior would be different?

- If "unix virtual terminals" and "windows cmd" are the two main text-based interfaces an app might use, where does powershell fall? Does it accommodate one or both of those two? Does it add a third object-based interface?

**Comment by mms710** [comment]
Getting back with an update on this. The first priority for the dev-infrastructure team is to stabilize our current CI stack (Homu/Buildbot) and support a Windows build. The second priority for dev-infra is to rebuild our CI stack so it's more scalable from a long-term perspective. Those two priorities will take almost all the team's time for the next few months. That being said, if there are some changes you all would like to request to make your lives easier while working on this, please don't hesitate to let me know. I'll work with the team to try to get it incorporated "on the side" as we push hard to get other priorities done.

**Comment by nuttycom** [comment]
At this point, our plan is to investigate a `nix`-based build system instead.
